### PR TITLE
Directionality formatter to move whitespace bidi handling into framework

### PR DESCRIPTION
**Pull request by GaryQian**
## Description

Mini-designdoc: go/flutter-rtl-ltr-whitespace-formatter-plan

Reproduced from the doc:

> According to https://www.w3.org/International/articles/inline-bidi-markup/uba-basics#neutralcharacters, Unicode Bidi will treat neutral characters as having directionality of the codepoints that surround it if it is surrounded by similarly directional codepoints, or otherwise the neutral space will take the directionality of the underlying paragraph.
> 
> This produces a very unpleasant typing experience that causes the caret to jump back and forth when typing spaces in a mixed RTL/LTR scenario. https://github.com/flutter/flutter/issues/34538 
> 
> We propose to move Flutter’s hack https://github.com/flutter/engine/pull/9489 out of layout (to remain faithful to spec) and into the framework as a textfield formatter.
> 
> The formatter would look for neutral whitespace runs that are surrounded by non-matching directionality text. We would insert an invisible RTL or LTR codepoint (see https://master-api.flutter.dev/flutter/foundation/Unicode/RLM-constant.html)  after the whitespace that matches the directionality of the preceding text.

The formater can either use Unicode ICU itself to determine directionality (which would be significantly more work and API surface), or, in this implementation, use regex to cover known sets of RTL characters.

The reason behind this change is that SkParagraph does not wish to implement a custom hack to achieve this behavior, a sentiment that we agree with. This level of preprocessing should exist in the framework as it is Flutter-specific. This behavior can reasonably be removed from LibTxt as well once this lands.

See the new test to see how this is expected to behave.

## Related Issues

https://github.com/flutter/flutter/issues/39420

## Tests

Test added to verify behavior of the formatter.

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [ ] All existing and new tests are passing.
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Did any tests fail when you ran them? Please read [Handling breaking changes].

- [x] No, no existing tests failed, so this is *not* a breaking change.

### CODE REVIEW

**Comment by GaryQian** [comment]
cc @Rusino @jason-simmons

File: packages/flutter/lib/src/painting/text_painter.dart
Comment (GaryQian) [inline packages/flutter/lib/src/painting/text_painter.dart]: Directionality markers take up zero space. This makes sure the range covers at least one real glyph when computing carets when it encounters a directionality marker

File: packages/flutter/lib/src/widgets/editable_text.dart
Comment (GaryQian) [inline packages/flutter/lib/src/widgets/editable_text.dart]: This probably needs a condition to not always run.

File: packages/flutter/lib/src/painting/text_painter.dart
Comment (xster) [inline packages/flutter/lib/src/painting/text_painter.dart]: code comment for what the scenario is and what this does

File: packages/flutter/lib/src/widgets/editable_text.dart
Comment (xster) [inline packages/flutter/lib/src/widgets/editable_text.dart]: The first sentence isn't clear to me after reading. "un-enclosed whitespace gets the directionality of the paragraph"  sounds like that's a generally desired behavior. "When typing in a direction that opposes the base direction of the paragraph" maybe describe how you can cause this.

File: packages/flutter/lib/src/widgets/editable_text.dart
Comment (xster) [inline packages/flutter/lib/src/widgets/editable_text.dart]: How did you decide whether user formatters should run before or after the correction?

File: packages/flutter/lib/src/widgets/editable_text.dart
Comment (xster) [inline packages/flutter/lib/src/widgets/editable_text.dart]: From the platform's perspective, 2 "current editing values" have been set right? Once with what the user typed and once with the re-formatted values. Could there be other side effects? For instance, could the keyboard's word completion suggestions flicker between 2 values depending on whether you're showing suggestions for the word on the left side or the right side?

File: packages/flutter/lib/src/widgets/editable_text.dart
Comment (xster) [inline packages/flutter/lib/src/widgets/editable_text.dart]: Do we have text edit benchmarks? This is adding an operation to all text entry right? Is there a cheaper short-out we can add anywhere in case there are no mixed directionalities anywhere?

File: packages/flutter/lib/src/widgets/editable_text.dart
Comment (xster) [inline packages/flutter/lib/src/widgets/editable_text.dart]: Also, add another set of tests to check that no other currently untested side effects occur when typing exclusively RTL or LTR text (i.e. _updateRemoteEditingValueIfNeeded doesn't get called a bunch of times etc)

File: packages/flutter/lib/src/widgets/editable_text.dart
Comment (xster) [inline packages/flutter/lib/src/widgets/editable_text.dart]: Don't abbreviate. Add code comment for what this is used for.

File: packages/flutter/lib/src/widgets/editable_text.dart
Comment (xster) [inline packages/flutter/lib/src/widgets/editable_text.dart]: Also describe what the content glyphs are supposed to be since the next maintainers can't read this and/or know what's the suitable criteria to maintain this / add / remove items.

File: packages/flutter/lib/src/widgets/editable_text.dart
Comment (xster) [inline packages/flutter/lib/src/widgets/editable_text.dart]: If you have this comment, you should also explain why you're not just importing those consts instead.

File: packages/flutter/lib/src/widgets/editable_text.dart
Comment (xster) [inline packages/flutter/lib/src/widgets/editable_text.dart]: Ah, you did it here. Can you benchmark the cost of this regex?

File: packages/flutter/lib/src/widgets/editable_text.dart
Comment (xster) [inline packages/flutter/lib/src/widgets/editable_text.dart]: rename or add comments to make the distinction between `out` and `newValue` clearer.

File: packages/flutter/test/widgets/editable_text_test.dart
Comment (xster) [inline packages/flutter/test/widgets/editable_text_test.dart]: these should all be separate tests (so debugger can spend less time figuring out what the test is without reading the implementation)

**Comment by xster** [comment]
The cirrus failure may be related

File: packages/flutter/lib/src/widgets/editable_text.dart
Comment (GaryQian) [reply packages/flutter/lib/src/widgets/editable_text.dart]: Some of these are being "premerged" from https://github.com/flutter/flutter/pull/52055 as I'm working on them in parallel. I expect the other patch to land first before this one.

Comment (GaryQian) [reply packages/flutter/lib/src/widgets/editable_text.dart]: This behavior should be handled in my work on the other patch, and should be replicated here.

File: packages/flutter/lib/src/widgets/editable_text.dart
Comment (GaryQian) [reply packages/flutter/lib/src/widgets/editable_text.dart]: The reasoning for calling it after is that this would be less risky as the user's formatters always receive the raw untouched data. Any changes we make are strictly after at least one pass through any user formatters.

File: packages/flutter/lib/src/widgets/editable_text.dart
Comment (GaryQian) [reply packages/flutter/lib/src/widgets/editable_text.dart]: Yeah will explain. Those consts in `Unicode` are actually strings, so not really usable without parsing them.

File: packages/flutter/lib/src/widgets/editable_text.dart
Comment (GaryQian) [reply packages/flutter/lib/src/widgets/editable_text.dart]: I don't think a properly implemented system should experience this. This extra call to _updateRemoteEditingValueIfNeeded is an effort to sync the remote and local values to the same thing. If it changes, one of the versions will end up being cached in _value, which is treated as ground truth. Obviously, if the formatter was not deterministic or had some cached counter or something, then you could achieve flickering/runaway behavior, but that is just an improper impl.

File: packages/flutter/lib/src/widgets/editable_text.dart
Comment (GaryQian) [reply packages/flutter/lib/src/widgets/editable_text.dart]: (see below, we short out if no mixed directionality is present)

File: packages/flutter/lib/src/widgets/editable_text.dart
Comment (GaryQian) [reply packages/flutter/lib/src/widgets/editable_text.dart]: Don't have numbers but I expect this to just be an O(N) single pass. Will update with more solid numbers.

Comment (GaryQian) [reply packages/flutter/lib/src/widgets/editable_text.dart]: Ok, so ran both regexes on a string of 235million long simple english string on my desktop, clocked in at ~439ms, so ~215ms for one pass of the regex 235million long strings. Just under 1ms/million characters.

10million runs of both regexes on 20k strings took ~398ms.

File: packages/flutter/lib/src/widgets/editable_text.dart
Comment (xster) [reply packages/flutter/lib/src/widgets/editable_text.dart]: sounds reasonable

File: packages/flutter/test/widgets/editable_text_test.dart
Comment (xster) [inline packages/flutter/test/widgets/editable_text_test.dart]: some grammar issues in this text?

File: packages/flutter/test/widgets/editable_text_test.dart
Comment (xster) [inline packages/flutter/test/widgets/editable_text_test.dart]: Do we have a test for no 0x200F ever show up if there are no mixed directions etc?

File: packages/flutter/lib/src/widgets/editable_text.dart
Comment (xster) [reply packages/flutter/lib/src/widgets/editable_text.dart]: SG

File: packages/flutter/test/widgets/editable_text_test.dart
Comment (GaryQian) [reply packages/flutter/test/widgets/editable_text_test.dart]: Yeah, can add that!

File: packages/flutter/lib/src/widgets/editable_text.dart
Comment (xster) [reply packages/flutter/lib/src/widgets/editable_text.dart]: great, thanks for checking

**Review by xster — APPROVED** [review]
LGTM

**Comment by GaryQian** [comment]
Holding off on this since https://github.com/flutter/flutter/pull/52055 is reverted (https://github.com/flutter/flutter/pull/52274) will land when https://github.com/flutter/flutter/pull/52055 is landed again.

**Comment by GaryQian** [comment]
Reland of dependent PR: https://github.com/flutter/flutter/pull/52366

File: packages/flutter/lib/src/painting/text_painter.dart
Comment (StetHD) [reply packages/flutter/lib/src/painting/text_painter.dart]: Hi, I'm finishing my flutter app, but some bugs are preventing me to release it.
in textfield if you have a persian numbers almost it always buggy, you know in persian numbers are from left to right but characters are from right to left, when users input for example mobile number, it's getting write in the opposite direction and it is just happening in ANDRIOD, in iOS it is smooth and fine

**Pull request merged**
